### DISCUSSION TITLE
NPM distribution for use in JavaScript projects?

### DISCUSSION THREAD

**Discussion by TheSpyder** [miniscruff/changie#550] [Ideas] [answered]
[edit: see #561]

I was wondering if you'd consider adding an NPM distribution of changie to aid using it in JavaScript projects.

The JavaScript ecosystem, for better or worse, doesn't like installing system-level tools; even binary tooling [goes into `node_modules`](https://www.reddit.com/r/ProgrammerHumor/comments/6s0wov/heaviest_objects_in_the_universe/) and is then executed locally within the project.

For an example of a tool written in go that offers a NPM distribution, look at esbuild:
https://www.npmjs.com/package/esbuild
https://github.com/evanw/esbuild
https://github.com/evanw/esbuild/tree/main/npm/esbuild
https://github.com/evanw/esbuild/tree/main/npm/%40esbuild

**Comment by miniscruff** [comment]
If there was a super easy and automatic way of wrapping a binary in JS maybe, but it would have to be entirely unrelated to the core changie work. I would not want to have to update the JS code/tooling whenever I work on a feature or bugfix. I would say the same is true for any other language distributions like python or whatever.

**Reply by TheSpyder** [reply]
Absolutely. The JS involved doesn't wrap the binary, it just moves the correct platform binary into a predictable location for NPM to link to. This JS doesn't change for each release, it's written once and then it's done.

There are two ways to run an npm binary distribution:
- ship all platform binaries in one NPM package (much easier for small tools)
- set up what esbuild has done, a new package for each platform and rely on NPM's peer dependency conditions to only download the correct one.

If you're open to this, I or my team should be able to spare some time to help set it up.

**Reply by miniscruff** [reply, marked as answer]
Hmm, I am willing to do the work of setting up npm for publishing, if you are able to setup any of the js things and update the release pipeline for it. 

For example using changies replacement feature to update any version string and the GitHub release workflow.

Will need to add it as an install option in the docs as well.

I might have some extra requirements but it has been a very long time sense I published to npm so I can't remember them right now.

**Reply by TheSpyder** [reply]
I'm not familiar with go release tooling but I'm happy to help any way I can!

I did a bit of a detailed writeup on how one of my projects achieves binary NPM distribution, but then after some quick research I found articles specifically describing how to distribute go binaries on NPM 😂 

https://github.com/sanathkr/go-npm
https://blog.xendit.engineer/how-we-repurposed-npm-to-publish-and-distribute-our-go-binaries-for-internal-cli-23981b80911b

The first downloads the appropriate binary at install time, which is certainly easy but not ideal for anyone using changie in multiple projects (which I will be). The second takes `go-npm` and reworks it to use multiple binaries published in a single NPM package, while a little wasteful on size I think this is the best option for now.

The sample JavaScript file in the article is overkill for what it does but I can shrink that down when I contribute it. The key to the whole thing is publishing an NPM package with every supported binary in a folder and the JavaScript, set up as a `postinstall`, copies the appropriate binary into a stable location defined by a `bin` config reference.

> For example using changies replacement feature to update any version string and the GitHub release workflow.

Yes it should be quite simple to use replacement to update the version string in `package.json`.

> Will need to add it as an install option in the docs as well.

Once this is all set up, the process for using changie via NPM will be
* install `npm i -D changie`
* run `npx changie`

Although technically `npx changie` will also work without installing first.

**Reply by miniscruff** [reply]
I think that all sounds good, I am curious how you are going to get the binaries for the installer, maybe a goreleaser hook or something? But whatever you need to do should be fine with me, let me know if I can help but I don't know much about this stuff either.

**Reply by TheSpyder** [reply]
In my project, which uses a matrix github actions to build because OCaml doesn't have easy cross-compilation, it grabs the binaries from the delegate action runners directly 😂 
https://github.com/TheSpyder/bs-atdgen-generator/blob/b1d421004b8161642ae482e622eee8db3caaba8f/.github/workflows/main.yml#L134-L153

Strangely, although your release action claims to mark the executables as artifacts, they don't show up under the summary artifacts list.

Maybe I can just build something into the existing release workflow instead of making it a new workflow. That sounds harder to test, although the process is [very well established](https://docs.github.com/en/actions/publishing-packages/publishing-nodejs-packages#publishing-packages-to-the-npm-registry).

**Reply by miniscruff** [reply]
You can use goreleaser locally to generate the builds in the dist folder, I think it's like --dry or --skip-publish

**Reply by TheSpyder** [reply]
I have the project running locally now but I need some help understanding whether I can tweak goreleaser to do what I need or should I write some scripts around what it produces. I'm happy to move this to an issue if you'd rather not get into development weeds in a discussion.

It looks like I can add additional steps into the `release.yml` github action that publish to NPM. The executables `goreleaser` builds seem to be available while that script runs, which is perfect; I could reference them directly in a `package.json` file and publish completely as-is. I was thinking of publishing everything except the `386` builds:
* `dist/changie_darwin_arm64/changie`
* `dist/changie_windows_arm64/changie.exe`
* `dist/changie_linux_amd64_v1/changie`
* `dist/changie_darwin_amd64_v1/changie`
* `dist/changie_linux_arm64/changie`
* `dist/changie_windows_amd64_v1/changie.exe`

The goreleaser docs say [don't rely on these folder names](https://goreleaser.com/customization/builds/#a-note-about-folder-names-inside-dist), but unless amd64 suddenly adds a v2 this seems quite stable. There is a `dist/artifacts.json` I could use to pull out the filenames if you'd rather I do it the correct way.

It would be nice if I could add a second `archives` specification for goreleaser using `format: binary`, but it seems this only supports one archive.

Ultimately what I need is a stable reference to these executables. Once I have that, I can point the `package.json` and `postinstall` scripts at them and the feature is done 🙂

**Reply by miniscruff** [reply]
I have the project running locally now but I need some help understanding whether I can tweak goreleaser to do what I need or should I write some scripts around what it produces. I'm happy to move this to an issue if you'd rather not get into development weeds in a discussion.

Yea an issue would be better.

    It looks like I can add additional steps into the release.yml github action that publish to NPM. The executables goreleaser builds seem to be available while that script runs, which is perfect; I could reference them directly in a package.json file and publish completely as-is. I was thinking of publishing everything except the 386 builds:

    * `dist/changie_darwin_arm64/changie`
    * `dist/changie_windows_arm64/changie.exe`
    * `dist/changie_linux_amd64_v1/changie`
    * `dist/changie_darwin_amd64_v1/changie`
    * `dist/changie_linux_arm64/changie`
    * `dist/changie_windows_amd64_v1/changie.exe`

    The goreleaser docs say [don't rely on these folder names](https://goreleaser.com/customization/builds/#a-note-about-folder-names-inside-dist), but unless amd64 suddenly adds a v2 this seems quite stable. There is a dist/artifacts.json I could use to pull out the filenames if you'd rather I do it the correct way.

    It would be nice if I could add a second archives specification for goreleaser using format: binary, but it seems this only supports one archive.

    Ultimately what I need is a stable reference to these executables. Once I have that, I can point the package.json and postinstall scripts at them and the feature is done 🙂

Hmm, if the artifacts.json file helps you I think that might be the way to go, but will have to see when you make a PR.

I will need to add some secrets for you in github actions for it to actually run as well, but that is fine.

**Reply by TheSpyder** [reply]
Yes I've been assuming we will sort out the NPM details after you're happy with my code 🙂

I'll make an issue for further discussion and start working on a script to collect the files using artifact.json!
